File: stacks/impl/StackUsingLinkedList.java
package stacks.impl;

import stacks.IStack;

import java.util.LinkedList;

// TODO: refactor this with custom LinkedList
public class StackUsingLinkedList<T> implements IStack<T> {
    LinkedList<T> linkedList;

    public StackUsingLinkedList() {
        linkedList = new LinkedList<>();
    }

    @Override
    public void push(T item) {
        linkedList.push(item);
    }

    @Override
    public T pop() {
        return linkedList.pop();
    }

    @Override
    public T peek() {
        return linkedList.peek();
    }

    @Override
    public boolean isEmpty() {
        return linkedList.isEmpty();
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("[");
        for (int i = linkedList.size() - 1; i >= 0; i--) {
            sb.append(linkedList.get(i));
            if (i != 0) sb.append(", ");
        }
        sb.append("]");
        return sb.toString();
    }
}

File: src/main/kotlin/dev/mslalith/stacks/impl/StackUsingLinkedList.kt
package dev.mslalith.stacks.impl

import dev.mslalith.linkedlist.ListNode
import dev.mslalith.stacks.IStack

class StackUsingLinkedList : IStack {

    private var head: ListNode? = null

    override fun push(item: Int) {
        val newNode = ListNode(item)

        if (isEmpty()) {
            head = newNode
            return
        }

        newNode.next = head
        head = newNode
    }

    override fun pop(): Int? {
        if (isEmpty()) return null

        val item = head?.`val`
        head = head?.next
        return item
    }

    override fun peek(): Int? = if (isEmpty()) null else head?.`val`

    override fun isEmpty(): Boolean = head == null

    override fun toString(): String = buildString {
        var curr = head
        if (curr != null) {
            append(curr.`val`)
            curr = curr.next
            while (curr != null) {
                insert(0, curr.`val`.toString() + ", ")
                curr = curr.next
            }
        }
        insert(0, "[")
        append("]")
    }
}
